src/tbbmalloc/large_objects.cpp

```diff
@@ -1020,17 +1020,33 @@ void ExtMemoryPool::freeLargeObjectList(LargeMemoryBlock *head)
 
 bool ExtMemoryPool::softCachesCleanup()
 {
-    return loc.regularCleanup();
+    bool ret = false;
+    if (!softCachesCleanupInProgress.fetch_or(0x1)) {
+        ret = loc.regularCleanup();
+        softCachesCleanupInProgress = 0;
+    }
+    return ret;
 }
 
-bool ExtMemoryPool::hardCachesCleanup()
+bool ExtMemoryPool::hardCachesCleanup(bool wait)
 {
+    if (hardCachesCleanupInProgress.fetch_or(0x1)) {
```

Comment:
this condition is the other way around compared to `softCachesCleanup`. I found this confusing since the only difference is `!` at the beginning. I'd suggest refactoring these to have the same flow.